lib/Runtime/Library/JavascriptSet.h

```diff
@@ -11,15 +11,64 @@ namespace Js
     public:
         typedef MapOrSetDataNode<Var> SetDataNode;
         typedef MapOrSetDataList<Var> SetDataList;
-        typedef JsUtil::BaseDictionary<Var, SetDataNode*, Recycler, PowerOf2SizePolicy, SameValueZeroComparer> SetDataSet;
+        typedef JsUtil::BaseDictionary<Var, SetDataNode*, Recycler, PowerOf2SizePolicy, SameValueZeroComparer> ComplexVarDataSet;
+        typedef JsUtil::BaseDictionary<Var, SetDataNode*, Recycler> SimpleVarDataSet;
 
     private:
+        enum class SetKind : uint8
+        {
+            // An EmptySet is a set containing no elements
+            EmptySet,
+            // An IntSet is a set containing only int elements
+            //
+            // Adding any TaggedInt or JavascriptNumber that can be represented as a TaggedInt
+            // will succeed and be stored as an int32 in the set, EXCEPT for the value -1
+            // Adding any other value will cause the set to be promoted to a SimpleVarSet or ComplexVarSet,
+            // depending on the value being added
+            //
+            // Deleting any element will also cause the set to be promoted to a SimpleVarSet
```

Comment:
Why promote on deletion? Is there a problem with BVSparse::Clear?

Reply:
Never mind, you answered elsewhere.

---
In reply to: [174844530](https://github.com/Microsoft/ChakraCore/pull/4816#discussion_r174844530) [](ancestors = 174844530)